File: python/euler009.py
"""
A Pythagorean triplet is a set of three natural numbers, a < b < c, for which,
a^(2) + b^(2) = c^(2)

For example, 3^(2) + 4^(2) = 9 + 16 = 25 = 5^(2).

There exists exactly one Pythagorean triplet for which a + b + c = 1000.
Find the product abc.
"""
import logging
from euler import euler_problem

@euler_problem
def euler009(n:int|str) -> int:
    n = int(n)
    for c in range(n - 2, 2, -1):
        for b in range(n - c - 1, 0, -1):
            a = n - c - b
            if b <= a:
                break # Go to next value of c

            if a ** 2 + b ** 2 == c ** 2:
                logging.info(f'Found triplets {a}, {b}, {c}, which sum to {n}')
                return a * b * c

    raise ValueError(f'No pythagorean triplets sum to {n}')


if __name__ == '__main__':
    print(euler009(12))
    print(euler009(1000))

File: java8/src/Euler009.java
import framework.EulerProblem;

import java.util.stream.IntStream;

/**
 * A Pythagorean triplet is a set of three natural numbers, a < b < c, for which,
 * <p>
 * a^2 + b^2 = c^2
 * For example, 3^2 + 4^2 = 9 + 16 = 25 = 5^2.
 * <p>
 * There exists exactly one Pythagorean triplet for which a + b + c = 1000.
 * Find the product abc.
 */
public class Euler009 extends EulerProblem<Integer> {
    @Override
    public long solve(Integer sum) {
        // Notes:u
        // Since a^2 + b^2 = c^2 and a + b + c == sum
        // it follows that:
        // b = (sum^2 - 2 * sum * a) / (2 * sum - 2 * a)
        // The solution will be any a, b where a, b > 0 and both integers.
        return IntStream.range(1, sum)
                .filter(a -> ((sum * sum) - 2 * sum * a) % (2 * sum - 2 * a) == 0)
                .map(a -> {
                    int b = ((sum * sum) - 2 * sum * a) / (2 * sum - 2 * a);
                    int c = sum - a - b;
                    return a * b * c;
                })
                .findFirst()
                .getAsInt();
    }

}